### PR TITLE
Adds piwik

### PR DESCRIPTION
**Pull request by pierreozoux**

### CODE REVIEW

**Comment by pierreozoux** [comment]
@tianon @yosifkit Do you have an idea? I have to admit, I don't undesrtand the error.

thanks!

**Comment by knickers** [comment]
It looks like the `logs` directory doesn't exist in bashbrew. Is it travis or piwik that's trying to write logs?

**Comment by md5** [comment]
@pierreozoux The library file format is wrong. You're missing the colons after the tag names.

**Comment by knickers** [comment]
Wow, I was way off. Thanks for not chewing me out.

**Comment by md5** [comment]
@knickers Of course not! :+1: 

The line that's supposed to create the log dir is here: https://github.com/docker-library/official-images/blob/50a535d8616faa4775527f2edf3c8a560007cdb4/bashbrew/bashbrew.sh#L153

The `touch` that was failing is here: https://github.com/docker-library/official-images/blob/50a535d8616faa4775527f2edf3c8a560007cdb4/bashbrew/bashbrew.sh#L327-L328

Since there wasn't a colon followed by a space in the library file line, `$repoTag` had the value `piwik:2.15.0 git://github.com/piwik/docker-piwik.git@21d998b0fee6e3b5155c6e0762f9e3e63ccd0f9a` (i.e. the whole line).

**Comment by pierreozoux** [comment]
Arf, really sorry, I checked 3 times, didn't see it. Now corrected.

**Comment by pierreozoux** [comment]
Do you need more info on my side to merge this?

**Comment by yosifkit** [comment]
Is email required for piwik to run?  The configuration included seems to be specific to one email setup.  So far, all of the official images do not include email "out-of-the-box", but let users setup their own email configuration.

**Comment by pierreozoux** [comment]
this is my personal view:
I think email is required to run piwik, like it is required to run ownCloud or WordPress. (If you forgot your password for instance, it will send you a reset email).
Plus I believe that the official images should include email, "out-of-the-box".
In this particular case, I find it an elegant way to ship email and is not that heavy (but I agree, it might not be the taste of everybody)

As a piwik image maintainer, if you prefer not having email, that's fine for me, I'll remove it from the setup (and add my email "layer" on top of the official image).

Let me know, and I'll do the necessary.

**Comment by tianon** [comment]
I agree at the high-level, but the devil is in the details when it comes to
"email, out-of-the-box" -- there's not really a great configuration that we
can even have a reasonably high degree of confidence is going to work for a
majority of users, so I'd much rather leave the entrypoint logic out and
document different ways to configure email via modifying ssmtp.conf (or
even better, link to ssmtp upstream's documentation).

**Comment by pierreozoux** [comment]
Ok, I'll update the PR.

**Comment by tianon** [comment]
Thanks :heart: !

**Comment by pierreozoux** [comment]
@tianon can I keep the ssmtp bin and remove the ssmtp configuration part in the entrypoint?

**Comment by tianon** [comment]
Yeah, I'm totally +1 to that -- I just don't like constraining the possible
ssmtp configuration. :+1: :metal:

**Comment by pierreozoux** [comment]
Ok, I finally found time \o/? How does it look like?

There is still https://github.com/piwik/docker-piwik/blob/master/docker-entrypoint.sh#L9
But if not, I'll have to have another image for piwik repo than the official one, based on the official one...

**Comment by yosifkit** [comment]
LGTM, that seems fine to me! The `chfn` could be in the Dockerfile, but either ways seems good.

Ping @tianon. (I'll start a build test in a few)

**Comment by pierreozoux** [comment]
@yosifkit arf, you;re right :)
Done!

**Comment by yosifkit** [comment]
LGTM, Build test of #1203; adf3e8f (`piwik`):

```console
$ bashbrew build "piwik"
Cloning piwik (git://github.com/piwik/docker-piwik.git) ...
Processing piwik:2.15.0 ...
Processing piwik:2.15 ...
Processing piwik:2 ...
Processing piwik:latest ...
$ bashbrew list --uniq "$url" | xargs test/run.sh
testing piwik:2.15.0
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed
```

**Comment by psftw** [comment]
This redistributes a GeoIP database snapshot, which has attribution requirements, per http://dev.maxmind.com/geoip/legacy/geolite/ .  I didn't see this mentioned on the Piwik website.  It should also be mentioned in the Docker image docs.

Waiting to check out the docs PR, but otherwise happy with the image.

**Comment by pierreozoux** [comment]
Here is the docs PR with the mention to GeoIP (good catch :) )

**Comment by tianon** [comment]
Nice catch, @psftw :heart:

I'm happy with where this is at now; LGTM :+1:

**Comment by yosifkit** [comment]
Build test hasn't changed; so we'll be good to merge once we are ready on the docs.

**Comment by pierreozoux** [comment]
Ok, I removed the docker-compose file.
Here is the change since last time.
https://github.com/piwik/docker-piwik/compare/e9656f5846e72ae180263a0f6c217b948b7b97a1...master

The biggest change is this commit: https://github.com/piwik/docker-piwik/commit/b2aed0b5e124eef8ff80d169db990b21e47b57e7

Before I was using a separate container for crontab. I didn't like it for the following reasons:
 - impossible to have the output (send to stdout, cron will want to email it)
 - cron is a bit always dark magic
 - a bit overkilling

With the bash loop in the background, logs are sent to stdout \o/

I hope this time is the good one!

**Comment by tianon** [comment]
I think the image is good, minus that new "cron" bit.

My problems with that are somewhat multi-faceted:
- it's not following best practice (now we have a spurious bash process running in the background)
- it's fragile -- if it fails at any point, it will just stop running (and unless the operator is paying attention, they won't even know that it's happened)
- it's trivial to accomplish with an external process, and by that I don't necessarily mean `cron` running in a second container, but I mean either doing `docker run ...` or `docker exec ...` via the system `cron` or even Jenkins (which is where I'd personally prefer to run such tasks)

So in summary, I'd like to see that bit added to the `docs` as either a `docker run ...` or `docker exec ...` line (maybe document both so they can choose what fits their needs best?) that folks should be encouraged to add to some other periodic tool such as a `cron` container, their system `cron`, or a CI instance with a recommended frequency (every 3600 seconds / hour like you've got there?)

**Comment by pierreozoux** [comment]
Ok, I totally removed the cron part.
I agree with you.

On top, by default, piwik ships with a web cron \o/

The only missing bit is the doc, I have troubles with the CI, help would
be appreciated :)

On 26-02-2016 17:52, Tianon Gravi wrote:
> I think the image is good, minus that new "cron" bit.
> 
> My problems with that are somewhat multi-faceted:
> - it's not following best practice (now we have a spurious bash process running in the background)
> - it's fragile -- if it fails at any point, it will just stop running (and unless the operator is paying attention, they won't even know that it's happened)
> - it's trivial to accomplish with an external process, and by that I don't necessarily mean `cron` running in a second container, but I mean either doing `docker run ...` or `docker exec ...` via the system `cron` or even Jenkins (which is where I'd personally prefer to run such tasks)
> 
> So in summary, I'd like to see that bit added to the `docs` as either a `docker run ...` or `docker exec ...` line (maybe document both so they can choose what fits their needs best?) that folks should be encouraged to add to some other periodic tool such as a `cron` container, their system `cron`, or a CI instance with a recommended frequency (every 3600 seconds / hour like you've got there?)
> 
> ---
> Reply to this email directly or view it on GitHub:
> https://github.com/docker-library/official-images/pull/1203#issuecomment-189392751
> 

-- 
I use PGP to protect our privacy, if you want to know more, you can
follow this
https://emailselfdefense.fsf.org/en/

If you have further questions, please do not hesitate to ask.
You can verify my public key here: https://keybase.io/pierreozoux

**Comment by tianon** [comment]
Added my final review comments in PR form at https://github.com/piwik/docker-piwik/pull/21, but other than that and updating the commit references here, I'm happy with where this is at!  Thanks for sticking with it.

**Comment by pierreozoux** [comment]
Ok!

I think we are ready :)

Thanks for your precious help!

**Comment by tianon** [comment]
LGTM

Build test of #1203; cffe0fc (`piwik`):

```console
$ bashbrew build "piwik"
Cloning piwik (git://github.com/piwik/docker-piwik.git) ...
Processing piwik:2.16.0 ...
Processing piwik:2.16 ...
Processing piwik:2 ...
Processing piwik:latest ...
$ bashbrew list --uniq "$url" | xargs test/run.sh
testing piwik:2.16.0
	'utc' [1/4]...passed
	'cve-2014--shellshock' [2/4]...passed
	'no-hard-coded-passwords' [3/4]...passed
	'override-cmd' [4/4]...passed
```

**Comment by yosifkit** [comment]
LGTM
